libr/core/cmd_open.c

```diff
@@ -1216,6 +1225,76 @@ static bool desc_list_json_cb(void *user, void *data, ut32 id) {
 	return true;
 }
 
+typedef struct {
+	int fd; // input and output
+	int inc;
+	int prevfd;
+	bool picknext;
+	bool found;
+	RCore *core;
+} RCmdOpData;
+
+static bool try_next(void *_user, void *_data, ut32 id) {
+	RCmdOpData *data = _user;
+	if (data->picknext) {
+		data->fd = id;
+		goto pickme;
+	}
+	if (id == data->fd) {
+		if (data->inc > 0) {
+			// pick next
+			data->picknext = true;
+		} else if (data->inc < 0) {
+			// pick prevfd
+			data->fd = data->prevfd;
+			goto pickme;
+		}
+	}
+	data->prevfd = id;
+	return true;
+pickme:
+	data->picknext = false;
+	data->found = true;
+	return false;
+	
+}
+
+static bool cmd_op(RCore *core, char mode, int fd) {
```

Comment:
this whole function and the one before looks a bit messy :/ I have an hard time following they are doing even if I know it.
I think something like this is clearer in its intention.
```C
static bool find_and_switch_fd(RCore *core, int mode) {
	RList *descs = r_io_desc_list (core->io);
	RIODesc *cur_desc = r_io_desc_get (core->io, r_io_fd_get_current (core->io));
	RListIter *it;
	RIODesc *desc;
	r_list_foreach (descs, it, desc) {
		if (desc == cur_desc) {
			RListIter *target = get_target_switch_fd (descs, it, mode);
			if (target) {
				RIODesc *set_desc = (RIODesc *)target->data;
				switch_fd (core, set_desc->fd);
				r_list_free (descs);
				return true;
			}
		}
	}
	r_list_free (descs);
	return false;
}
```

Reply:
My implementation won't need to iterate over all descs in all the cases, also your implementation is not handling the case an fd is passed. I agree that the code is messy, and i consider it wip, but i would prefer to not add new apis to r_io when we can iterate directly. I agree the ridstorage usage is not as handy as r_list. I'll try to refactor this a bit more when i get some spare time

Reply:
it doesn't need to handle the case when the fd is passed, because that one can be done just with `switch_fd (core, fd)`.

Reply:
How can you avoid iterating over all the descs up to the current one? I did not find any API to directly get the previous/next desc by just having the current one without the list (or without iterating with the foreach).

Also, why so many problems to add an API to r_io? It is just an addition and a one that may come handy in more than one case. If it was a change, I would have understood, but a common addition like the "to_list" method seems nice to me.

Reply:
You can also iterate over r_id_storage without passing a function-pointer. If you want I can create a makro for that. So far I haven't seen necessity for that as I personally find passing fcn-pointers leads to shorter functions. no need for to_list imo.

Reply:
you can use something like this to find the next fd (I hope I understodd correctly, that this is what you want to do)
```
ut32 next_fd = cur_fd + 1;
RIODesc *next = NULL;
for (; next_fd != cur_fd; next_fd = (next_fd + 1) % io->files->size) {
	next = r_io_desc_get (io, next_fd);
	if (next) {
		break;
	}
}
```

I think having a fcn in io and id_storage for this would be good. I can do that change for you @trufae today/evening